### ISSUE TITLE
Shoud a \RuntimeException be considered outside to your namespace?

### ISSUE THREAD

**Issue by ricfio** [phparkitect/arkitect#200]
### Bug Report

|    Q            |   A
|---------------- | ------
| BC Break        | no
| Library Version | 0.2.4 
| PHP version     | 8.0.12

#### Summary

I'm not sure that this is a bug, it could be only a configuration issue...
Shoud a \RuntimeException be considered outside to your namespace `App\Domain` (using `ResideInOneOfTheseNamespaces`) ?

#### How to reproduce

Use the following rule to check a class in your `App\Domain` that throws a RuntimeException in any method:
```
    $rules[] = Rule::allClasses()
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new NotHaveDependencyOutsideNamespace('App\Domain'))
        ->because('we want protect our domain');
```

#### Expected behavior

I would like configure phparkitect to ignore some namespaces such as the root namespace (for example for the class: \RuntimeException) when checking some rules.

**Comment by ricfio** [comment]
A better chance would be to express a more specific configuration rule:
```
    $rules[] = Rule::allClasses()
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new NotHaveDependencyOutsideNamespace('App\Domain'))
        ->because('we want protect our domain');
```

changing `new NotHaveDependencyOutsideNamespace('App\Domain')` in something like this (for example):
```
    new NotHaveDependencyOutsideNamespace('App\Domain')
    ->whenDependencyIsInsideNamespace('App')
```

**Comment by micheleorselli** [comment]
TLDR: depending on classes in the global namespace is something you probably want to do, for this reason I am for ignoring this kind of dependencies.

The Long version: 
Based on my (biased) experience you usually want to avoid a dependency on some specific classes (like `\RuntimeException` for instance) in the global namespace. For this use case a rule like `NotDependencyOnClasses` could be useful 

```
$rules[] = Rule::allClasses()
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new NotDependOnClasses('\RuntimeException', '\DomainException'))
        ->because('we to use custom exception and avoid general ones');
```

**Comment by ricfio** [comment]
Your experience is very helpful to me, and I have agree with you about the Exceptions, even in my opinion is better have specialized Exceptions into the my Domain, for this reason I'm already working on this...

That said, however, I have this problem on any type of class in global namespace... so should I use specialized classes or interface also for classes/interfaces such as DateTimeImmultable, Stringable, etc. ?

Could you suggest me the better choice in your opinion, please?

**Comment by AlessandroMinoccheri** [comment]
At the moment there is another trick for this situation that we have discussed in past in addition to the one proposed by @micheleorselli.

Instead of using:

```
->should(new NotHaveDependencyOutsideNamespace('App\Domain'))
```

you can use:

```
->should(new DependsOnlyOnTheseNamespaces('App\Domain', 'OtherNamespace\'))
```

You can try if the root namespace could be ok for this situation.

What do you think @ricfio?

**Comment by ricfio** [comment]
Hm, ok this seems to work:
```
        ...
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new DependsOnlyOnTheseNamespaces('App\Domain', '\\'))
        ...
```

However If I remove 'App\Domain' from the DependsOnlyOnTheseNamespaces constructor parameters (as follow) there are errors:
```
        ...
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new DependsOnlyOnTheseNamespaces('App\Domain', '\\'))
        ...
```

In my opinion the domain used in ResideInOneOfTheseNamespaces constructor parameters could be implicit.

**Comment by ricfio** [comment]
> Even If I implements specialized Exceptions, extending (for example) \RuntimeException, these classes will be inside my domain namespace (App\Domain), so the check rule will be violated again.

Could anyone reply me on this? Suggestions ?
Thank you.

**Comment by micheleorselli** [comment]
Hi @ricfio, 

> Even If I implements specialized Exceptions, extending (for example) \RuntimeException, these classes will be inside my domain namespace (App\Domain), so the check rule will be violated again.

Well, the rule will be violated for the same reason above, because your code actually depends on \RuntimeException. Does  @AlessandroMinoccheri's suggestion covers this case?

**Comment by micheleorselli** [comment]
> Hm, ok this seems to work:
> 
> ```
>         ...
>         ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
>         ->should(new DependsOnlyOnTheseNamespaces('App\Domain', '\\'))
>         ...
> ```
> 
> However If I remove 'App\Domain' from the DependsOnlyOnTheseNamespaces constructor parameters (as follow) there are errors:
> 
> ```
>         ...
>         ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
>         ->should(new DependsOnlyOnTheseNamespaces('App\Domain', '\\'))
>         ...
> ```
> 
> In my opinion the namespace in the ResideInOneOfTheseNamespaces constructor parameters ('App\Domain') should be implicit.

Would you mind explain it again? The 2 code snippets are the same to me, are there differences I am not seeing?

**Comment by ricfio** [comment]
Ops, I have edited my previous comment... here the 2nd snippet (edited):
```
        ...
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new DependsOnlyOnTheseNamespaces('\\'))
        ...
```

**Comment by micheleorselli** [comment]
Oh, now I get what you mean... well I guess it makes sense to fix this behaviour. I do not have a lot of time to look into ti right now, feel free to send a PR :-)

**Comment by AlessandroMinoccheri** [comment]
@ricfio I understand what you mean, in my opinion, it's better this:

```
        ->that(new ResideInOneOfTheseNamespaces('App\Domain'))
        ->should(new DependsOnlyOnTheseNamespaces('App\Domain', '\\'))
```

Because is more explicit when you read the rule, I know that it could be more verbose and you can omit `'App\Domain', but in terms of readability for me is better.

**Comment by micheleorselli** [comment]
@AlessandroMinoccheri @fain182 well, I guess it is safe to assume that a class can depend on classes on the same namespace. If that is true, we can safely omit the starting namespace... what do you guys think?

**Comment by fain182** [comment]
Yes, it is a good default imho.
